### ISSUE TITLE
Performance issues with latest minor versions `colinmollenhour/cache-backend-redis` and `colinmollenhour/credis` 

### ISSUE THREAD

**Issue by glo71317** [colinmollenhour/Cm_Cache_Backend_Redis#181]
Hi,

When we ran the `composer update` then minor version is updated of `colinmollenhour/cache-backend-redis` and `colinmollenhour/credis` then we received the large number of performance issues in adobe commerce.

So, during investigation we found these are dependency which was causing issue so we restricted these dependency with older version of both the packages and it worked well to us.

Still more investigation is pending, right now release is very near so we decided to unblock the delivery for upcoming release. will continue investigating once we will get time from current release.

You can also revisit the changes which are introduced in latest version for those packages which i mentioned above.
Find the log
[report.zip](https://github.com/colinmollenhour/Cm_Cache_Backend_Redis/files/14121835/report.zip)

For more detail -  https://github.com/magento/magento2/commit/9369884f8c40745291347470c5910a0f01a556af#commitcomment-137202896

**Comment by Xon** [comment]
The lock file indicates you went back from v1.17.1 => v1.16.0

The major performance related change was for lua to be enabled by default in v1.17.0

**Comment by Xon** [comment]
If anything the performance impact should be from disabling lua, The lockfile went from having lua enabled to lua disabled due to the version change

**Comment by glo71317** [comment]
@colinmollenhour You might be right, lua should not be enabled in minor version, best way to introduce Backward compatibility issues in major version. Normally restrict to minor or patch version of any library is not good way to resolve the issue.

Right now it is time consuming to pin point the issue in adobe commerce. Have to see how to proceed in this case.

**Comment by colinmollenhour** [comment]
I didn't consider this a backward-compatibility issue as the lua mode has been present for a very long time and no code APIs changed. It is unexpected that it had a noticeable negative impact on performance, though. It was released as a minor version update and not a patch update so you could lock your version to `~1.16.0` allowing for future patch updates while not upgrading to 1.17.0.

**Comment by hostep** [comment]
It appears Magento 2.4.7 will come with a configuration option to enable or disable lua: https://github.com/magento/magento2/commit/b5ede695df511eb42e5a93969e29fdbb3d02eb77

**Comment by colinmollenhour** [comment]
Thanks for the update, @hostep 

@glo71317 I'd still appreciate any insight into under what situations the lua flag is a net negative if you have any more details. If the read:write ratio is "normal" compared to what I've seen in my experience the impact should be minimal so I suspect it is unique to the test environment, but I've been wrong before. :)

**Comment by glo71317** [comment]
@colinmollenhour As we had adapted the lua changes at our end, By default lua will be false in adobe commerce when redis is enabled if someone want use the lua flag enabled then end user can make it enabled lua flag and continue to achieve their requirements.
Thanks for the update!

**Comment by hostep** [comment]
@glo71317: I agree with Colin in that it would be a good idea to dig a bit deeper in Magento's cache usage, to figure out exactly what is the root cause of this problem instead of hiding it with a flag.

**Comment by glo71317** [comment]
@hostep we are not hiding anything will speak to PO for investigating more bit in deep to share read:write ratio. But surely will do after current release finish.

**Comment by igorwulff** [comment]
@hostep  We have seen similar issues with Redis after use_lua was by default set to 1. We had to set it to 0 again, to stabilize Redis read timeout errors and other errors from popping up. The backend_clean_cache cron from Magento also started to fail after use_lua=1 was set by default.

**Comment by Nuranto** [comment]
maybe related : https://github.com/magento/magento2/issues/38668

**Comment by damienwebdev** [comment]
I think I have a valid reproduction of this on Magento b2.4.6-p6 and colinmollenhour/cache-backend-redis:v1.17.1

My redis monitor is full of:

> 1723139520.164118 [0 IP:PORT] "evalsha" "1617c9fb2bda7d790bb1aaa320c1099d81825e64" "3" "40d_CONFIG_SCOPES" "40d_CONFIG" "40d_MAGE" "zc:k:" "d" "t" "m" "i" "zc:tags" "zc:ti:" "zc:ids" "40d_SYSTEM" "gz:..." "40d_CONFIG_SCOPES,40d_CONFIG,40d_MAGE" "1723139520" "1" "" "0"
1723139520.778961 [0 lua] "HMSET" "zc:k:40d_SYSTEM" "d" "gz:..." "t" "40d_CONFIG_SCOPES,40d_CONFIG,40d_MAGE" "m" "1723139520" "i" "1"

Upon repeated attempts to make cacheable GET requests. I would assume that the contents of the cache contain something that causes a hash to be mismatched for the same content over and over.

I'm going to try to dig further to see what (in these cache keys) is causing a miss.

**Comment by colinmollenhour** [comment]
Thanks for the info, @damienwebdev 

That's the `save()` command for the 'SYSTEM' key. Are all the updates referencing the '40d_SYSTEM' value in the same position or is that value not repeated at all or rarely?

The Magento config (which consists of many sections/scopes each with their own key) should be treated as basically immutable, there should be absolutely no code that is changing it directly; only a user-triggered config update or new code deployment should touch the config. I've seen extensions that do very naughty things like using the config to store state data for frequently updated state like after a cron job runs - this is what `core/flag` is good for (in M1, not sure what the M2 equivalent is called). Also, it is bad to use "Flush" on a busy site as it is possible to cause a "stampede" of updates (this was fixed thoroughly in OpenMage, not sure about other variants). It's even worse when there is a config flush buried in an extension somewhere..

If users are updating the config frequently, it is true that it is not a well-optimized scenario as Magento re-saves the entire config even if only a single value is changed so this too should be avoided. E.g. perhaps some config data should instead live in separate tables and not the config. This really applies to any backend, but the Redis lua mode *does* sacrifice some save performance for stability - without it there is some possibility for race conditions since it requires multiple updates. In general, the assumption is that saves are rare and reads are not, so if saves are not rare, then lua mode will not be a good candidate.

If the core uses a pattern that needlessly causes lots of config updates, then I'd argue that should be fixed first.

Also, I've never checked this in M2 but in OpenMage the config uses an infinite lifetime to prevent the config from being evicted - only an intentional act should clear the config. Lastly, the "allkeys-random" eviction policy should definitely not be used.

However, as I've looked back on this, the current lua implementation does *everything* in one call but since it involves extra encoding/decoding steps and the config data can be quite large for some stores, it may make sense to split it into two calls: one that saves the key data (the first hmset) and then the evalsha to update all of the tags. I think only an implementation comparing benchmarks can answer that.. I'm not planning on tackling this at the moment, just to be clear.

**Comment by damienwebdev** [comment]
@colinmollenhour after digging into this further, it appears my issue is caused by missing the retry_reads_on_master and then triggering a cache clear. Even on a site with absolutely no volume, I was able to see this behavior on a system with 1 master and 9 replicas simply by loading pages in the Magento admin.

Once I added that config, my specific issue resolved. Apologies for the noise.

**Comment by colinmollenhour** [comment]
Ahh, that will do it, sounds like the replication latency caused a cache saving stampede. OpenMage fixed this in https://github.com/OpenMage/magento-lts/pull/3533/files even without `retry_reads_on_master` by using MySQL's `GET_LOCK`. I'm not sure what the M2 implementation is doing.

I tested the idea of moving the hmset that does the main data save outside of the lua script and it's surprisingly quite a bit slower.. So as of now I don't know of any way to improve either lua or non-lua modes.

**Comment by hostep** [comment]
Sorry, I don't understand how one can be a key and the other the value? That makes little sense to me?

**Comment by hostep** [comment]
A new change is coming in Magento (probably in version 2.4.8?) which may or may not have anything to do with this issue: https://github.com/magento/magento2/commit/28c0e735bf79742e510b47b4cf7f13761af4da3a

It seems like that confusing 2nd `_useLua` config flag is being phased out. And there is now a new `use_lua_on_gc` flag being introduced.
Not sure what all of this means (possibly what [Colin mentioned earlier about race conditions](https://github.com/colinmollenhour/Cm_Cache_Backend_Redis/issues/181#issuecomment-1960900908)?). No idea if this will have a further impact on performance somehow in Magento2.

**Comment by colinmollenhour** [comment]
Interesting, I guess they've determined lua mode is good in one case and not the other.
I don't think GC is very important to do often if at all unless your usage patterns have lots of temporary keys. But if it's mostly the same keys (like product ids) then GC is probably unnecessary.

**Comment by RicardsRikmanis** [comment]
Our team just upgraded to 2.4.8 and what we noticed is how Magento 2 `backend_clean_cache` cron job locked up Redis for couple of minutes due to the Lua GC. I'd rather live with some rare race conditions than Redis locking up with busy status.

**Comment by nemphys** [comment]
@colinmollenhour glad if I can help, below is the non-LUA implementation of the GC routine in javascript:

```
const MAX_CACHE_SIZE = 300_000;
const allTags = await redisClient.smembers(SET_TAGS);
const allTagsCount = allTags.length;
let exists = new Map();
let tagCounter = 0;
for (let tag of allTags) {
    const tagKey = PREFIX_TAG_IDS + tag;

    let cursor = '0';
    let totalNotExpired = 0;

    do {
        const [ nextCursor, members ] = await redisClient.sscan(tagKey, cursor, 'COUNT', 500);
        cursor = nextCursor;

        if (members.length === 0) continue;

        // Build list of IDs that are not in the global cache yet
        const toLookup = members.filter(m => !exists.has(m));

        // Pipeline EXISTS for uncached IDs
        let batchExists;
        if (toLookup.length > 0) {
            batchExists = new Map();
            const pipeline = redisClient.pipeline();
            for (const id of toLookup) {
                pipeline.exists(PREFIX_KEY + id);
            }
            const pipelineResults = await pipeline.exec();
            let resIndex = 0;
            for (const id of toLookup) {
                const [ , existsVal ] = pipelineResults[resIndex++];
                batchExists.set(id, existsVal);

                // Store in global cache while under limit
                if (exists.size < MAX_CACHE_SIZE) {
                    exists.set(id, existsVal);
                }
            }
        }

        // Now classify members as expired / non-expired
        const expiredIds = members.filter(m => (exists.has(m) ? exists.get(m) : batchExists?.get(m)) === 0);
        totalNotExpired += members.length - expiredIds.length;

        // Remove expired IDs in chunks
        if (expiredIds.length > 0) {
            const chunkSize = 100;
            for (let i = 0; i < expiredIds.length; i += chunkSize) {
                const slice = expiredIds.slice(i, i + chunkSize);
                const pipeline = redisClient.pipeline();
                for (const id of slice) {
                    // We don't use a plain srem here, since we need a second (atomic) existence check because the key might have been re-inserted in the meantime
                    pipeline.safeDeleteTagKey(
                        3,
                        PREFIX_KEY + id,
                        tagKey,
                        SET_IDS,
                        id,
                        config.notMatchingTags ? 1 : 0
                    );
                }
                await pipeline.exec();
            }
        }

    } while (cursor !== '0');

    // Delete tag entirely if nothing remains
    if (totalNotExpired === 0) {
        let multi = redisClient.multi();
        multi.del(tagKey);
        multi.srem(SET_TAGS, tag);
        await multi.exec();
    }
}
exists.clear();
```

and the safeDeleteTagKey LUA script:

```
local existsNow = redis.call('EXISTS', KEYS[1])
if existsNow == 0 then
    redis.call('SREM', KEYS[2], ARGV[1])
    if ARGV[2] == '1' then
        redis.call('SREM', KEYS[3], ARGV[1])
    end
end
return existsNow
```

**Comment by damienwebdev** [comment]
@colinmollenhour I'm fairly confident I have a replication of this at least for the Magento side. We recently upgraded a client application from v2.4.6 to v2.4.7. Why this occurs as a result of the upgrade is unclear to me as `colinmollenhour/cache-backend-redis` and `colinmollenhour/credis` are the same versions. That said, I'm currently of the opinion this is a bug [in this lua script](https://github.com/colinmollenhour/Cm_Cache_Backend_Redis/blob/master/Cm/Cache/Backend/Redis.php#L613) in `colinmollenhour/cache-backend-redis`. Unfortunately, I do not know what this code's intent is, so it's hard for me to say what's wrong.

## Original:
- Magento v2.4.6 (colinmollenhour/cache-backend-redis:1.17.1, colinmollenhour/credis:1.17.0)
- Magento v2.4.7 (colinmollenhour/cache-backend-redis:1.17.1, colinmollenhour/credis:1.17.0)

I haven't compared the yet, but the issue in my case is the following key:

`Magento_Framework_GraphQlSchemaStitching_Config_Data`

In particular, in my case, the issue appears to be that during save, the key is never actually saved.

In specific, when `$this->useLua` is true, during save it hits this branch

<img width="563" height="338" alt="Image" src="https://github.com/user-attachments/assets/43ef6e6d-6eba-481f-bc61-296b4d35483c" />

<img width="730" height="231" alt="Image" src="https://github.com/user-attachments/assets/a129bd38-8f82-4a6d-8b62-3c917ded69b7" />

And I wind up in a situation where `$this->_redis->sRem` is called several thousand times as it tries to clear tags.

At the same time (just for a comparison, when `$this->useLua` is false:

<img width="839" height="142" alt="Image" src="https://github.com/user-attachments/assets/0f1d138f-9ec6-417e-b894-9d97dd8dbe51" />

`$oldTags is empty for the same key.`

Do note that in this example, the `Magento_Framework_GraphQlSchemaStitching_Config_Data` is being saved for the very first time, its not in redis ahead of time.

**Comment by colinmollenhour** [comment]
I published [1.18.0 ](https://github.com/colinmollenhour/Cm_Cache_Backend_Redis/releases/tag/1.18.0) which fixes the save corruption and also uses the improved garbage collection (as well as adds Dragonfly support).
